### PR TITLE
Task unification todo task removal

### PR DESCRIPTION
**Pull request by jumasheff**

### CODE REVIEW

File: orchestra/migrations/0087_todo_task_data_migration.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
+# Generated by Django 2.2.13 on 2020-09-04 14:49
+
+from django.db import migrations
+from django.db.models import F
```
Comment (jumasheff) [inline orchestra/migrations/0087_todo_task_data_migration.py]: Remove `F` import

File: orchestra/migrations/0087_todo_task_data_migration.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+from django.db.models import OuterRef
+from django.db.models import Subquery
+
+
+def fill_project_and_step_fields(apps, schema_editor):
```
Comment (marcua) [inline orchestra/migrations/0087_todo_task_data_migration.py:9]: sanity-checking: does this code work, or is it still WIP?

Comment (marcua) [reply orchestra/migrations/0087_todo_task_data_migration.py:9]: if it does work, how long does it take to run? :)

File: orchestra/tests/test_todos.py [hunk 3]
```diff
@@ -25,13 +25,13 @@
...
 
-def _todo_data(task, title, completed,
+def _todo_data(task_id, title, completed,
                skipped_datetime=None, start_by=None,
                due=None, parent_todo=None, template=None,
                activity_log=str({'actions': []}), qa=None,
                project=None, step=None):
     return {
-        'task': task.id,
+        'task': task_id,
```
Comment (marcua) [inline orchestra/tests/test_todos.py]: should we just be eliminating task as an argument here entirely?

File: orchestra/tests/test_todos.py [hunk 4]
```diff
@@ -172,14 +174,16 @@ def test_todos_list_create(self):
 
     def test_todos_list_create_permissions(self):
         # Can't make requests for projects in which you're uninvolved.
-        task = TaskFactory()
+        project = ProjectFactory()
+        step = StepFactory()
+        task = TaskFactory(project=project, step=step)
```
Comment (marcua) [inline orchestra/tests/test_todos.py]: do we need a task at all? (it's possible we do)

File: orchestra/todos/auth.py [hunk 5]
```diff
@@ -34,29 +33,17 @@ class IsAssociatedWithProject(permissions.BasePermission):
 
     def has_permission(self, request, view):
         worker = Worker.objects.get(user=request.user)
```
Comment (marcua) [inline orchestra/todos/auth.py:41]: i liked in the old format that it was explicit that GET and POST got their arguments from different sources. what are some of the benefits of this new implicit approach?

File: orchestra/tests/test_todos.py (hunk 4 above)
Comment (jumasheff) [reply orchestra/tests/test_todos.py]: We didn't need it :)

File: orchestra/todos/auth.py (hunk 5 above)
Comment (jumasheff) [reply orchestra/todos/auth.py:41]: We now have PATCH and PUT requests. Besides, `query_params` is not exclusive to GET requests anymore –– our filtering logic is based on it and can potentially be used across all request types :)

File: orchestra/migrations/0087_todo_task_data_migration.py (hunk 2 above)
Comment (jumasheff) [reply orchestra/migrations/0087_todo_task_data_migration.py:9]: It takes 41 seconds :)

File: orchestra/migrations/0087_todo_task_data_migration.py [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+from django.db import migrations
+from django.db.models import OuterRef
+from django.db.models import Subquery
+
+
+def fill_project_and_step_fields(apps, schema_editor):
+    Step = apps.get_model('orchestra', 'Step')
+    Project = apps.get_model('orchestra', 'Project')
+    Todo = apps.get_model('orchestra', 'Todo')
+    o = OuterRef('id')
```
Comment (marcua) [inline orchestra/migrations/0087_todo_task_data_migration.py:14]: you'll have to explain this to me live :) specifically, why we need the [:1].

Comment (marcua) [reply orchestra/migrations/0087_todo_task_data_migration.py:14]: Perhaps a broader comment explaining what this code does will help.

File: orchestra/migrations/0087_todo_task_data_migration.py [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+    step_outerref_filter_qs = Step.objects.filter(tasks__todos__id=o)
+    project_outterref_filter_qs = Project.objects.filter(tasks__todos__id=o)
+    subq_step = step_outerref_filter_qs.values('id')[:1]
+    subq_proj = project_outterref_filter_qs.values('id')[:1]
+    Todo.objects.update(project=Subquery(subq_proj), step=Subquery(subq_step))
```
Comment (marcua) [inline orchestra/migrations/0087_todo_task_data_migration.py:23]: having some sort of testing/verification might be nice. like just at the shell, before the code runs, loop through all todos and show that the number of todos where the task.project == project and task.step == step is 0, and after the migration, the number that meet that criterion should be the number of todos.

Comment (marcua) [reply orchestra/migrations/0087_todo_task_data_migration.py:23]: (not code reviewed code, just verification pre-/post-migration.)

File: orchestra/models/core/models.py [hunk 8]
```diff
@@ -666,7 +666,8 @@ class Status(ChoicesEnum):
         DECLINED = 'declined'
 
     task = models.ForeignKey(Task, null=True, blank=True,
-                             related_name='todos', on_delete=models.SET_NULL)
+                             related_name='todos_old',
```
Comment (marcua) [inline orchestra/models/core/models.py]:
```suggestion
                             related_name='todos_deprecated',
```

Comment (marcua) [reply orchestra/models/core/models.py]: (plus update migration :))

File: orchestra/static/orchestra/todos/todo-list.directive.es6.js [hunk 9]
```diff
@@ -27,19 +27,22 @@ export default function todoList (orchestraApi) {
...
-        task: taskId,
-        title,
-        completed,
-        start_by_datetime: startDate,
-        due_datetime: dueDate
-      }).then((taskData) => {
-        todoList.todos.unshift(taskData)
-        return taskData
-      })
+      const createTodo = (title, completed, startDate, dueDate) => {
```
Comment (marcua) [inline orchestra/static/orchestra/todos/todo-list.directive.es6.js:29]: why did `taskId` get passed in, but now we'd get the step slug from `todoList.newTodoStepSlug`?

Comment (marcua) [reply orchestra/static/orchestra/todos/todo-list.directive.es6.js:29]: (not saying it's wrong, just making sure i understand)

File: orchestra/static/orchestra/todos/todo-list.directive.es6.js [hunk 10]
```diff
@@ -51,7 +54,7 @@ export default function todoList (orchestraApi) {
       }
 
       todoList.sendToPending = () => {
-        createTodo(todoList.taskId, 'Send task to pending state', true)
+        createTodo('Send task to pending state', true)
```
Comment (marcua) [inline orchestra/static/orchestra/todos/todo-list.directive.es6.js]: manual testing: does the pending button work?

File: orchestra/static/orchestra/todos/todo-list.directive.es6.js [hunk 11]
```diff
@@ -105,7 +108,7 @@ export default function todoList (orchestraApi) {
         activityLog['actions'].push({
           'action': action,
           'datetime': activityDatetime,
-          'step_slug': todoList.taskSlugs[todoList.taskId]
+          'step_slug': todo.step
```
Comment (marcua) [inline orchestra/static/orchestra/todos/todo-list.directive.es6.js:98]: do we still use `todoList.taskSlugs` anywhere?

File: orchestra/tests/test_todos.py [hunk 12]
```diff
@@ -67,10 +67,17 @@ def setUp(self):
...
-        self.step = StepFactory()
-        self.project = ProjectFactory()
+        self.workflow_version = WorkflowVersionFactory()
+        self.step = StepFactory(
+            slug='step-slug',
+            workflow_version=self.workflow_version)
+        self.project = ProjectFactory(
+            workflow_version=self.workflow_version)
         self.tasks = Task.objects.filter(assignments__worker=self.worker)
         self.task = self.tasks[0]
```
Comment (marcua) [inline orchestra/tests/test_todos.py]: do we still need self.task for anything?

File: orchestra/todos/api.py [hunk 13]
```diff
@@ -6,7 +6,8 @@
 
 from orchestra.models import TodoListTemplate
 from orchestra.models import Todo
-from orchestra.models import Task
+from orchestra.models import Project
+from orchestra.utils.view_helpers import get_step_by_project_id_and_step_slug
```
Comment (marcua) [inline orchestra/todos/api.py]: view_helpers is a funny module name in the sense that it implies its contents aren't usable elsewhere (not true---you use this function in serializers :))

File: orchestra/todos/auth.py (hunk 5 above)
Comment (marcua) [reply orchestra/todos/auth.py:41]: ok cool -> want to just explain in a comment why we look for project_id in so many places? :)

File: orchestra/todos/serializers.py [hunk 14]
```diff
@@ -70,26 +71,46 @@ def update(self, instances, validated_data):
...
+
 class BulkTodoSerializer(serializers.ModelSerializer):
     json_schemas = {
         'activity_log': TodoActionListSchema
     }
 
-    # TODO(murat): Remove this validation when step will be marked as required
+    step = StepField()
+
+    # TODO(murat): Remove this validation when project
```
Comment (marcua) [inline orchestra/todos/serializers.py:89]:
```suggestion
    # TODO(murat): Remove this validation when project 
    # becomes a required field in models.
```

File: orchestra/todos/serializers.py [hunk 15]
```diff
@@ -70,26 +71,46 @@ def update(self, instances, validated_data):
...
+    # will become a required field in models
```
Comment (marcua) [inline orchestra/todos/serializers.py]:
```suggestion
    # becomes a required field in models.
```

File: orchestra/todos/urls.py [hunk 16]
```diff
@@ -13,7 +13,8 @@
     url(r'^todo/$',
         TodoViewset.as_view({'post': 'create', 'get': 'list'}), name='todos'),
     url(r'^todo/(?P<pk>[0-9]+)/$',
-        TodoViewset.as_view({'put': 'update'}), name='todo'),
+        TodoViewset.as_view(
```
Comment (marcua) [inline orchestra/todos/urls.py]: sanity-checking what this is for?

File: orchestra/todos/views.py [hunk 17]
```diff
@@ -33,12 +33,12 @@
               logger=logger)
 def update_todos_from_todolist_template(request):
     todolist_template_slug = request.data.get('todolist_template')
-    task_id = request.data.get('task')
+    project_id = request.data.get('project')
+    step_slug = request.data.get('step')
```
Comment (marcua) [inline orchestra/todos/views.py:37]: is is step-slug, or does `_set_step_to_validated_data` make it the actual step?

File: orchestra/todos/views.py [hunk 18]
```diff
@@ -33,12 +33,12 @@
...
     try:
-        add_todolist_template(todolist_template_slug, task_id)
-        project = Task.objects.get(id=task_id).project
+        add_todolist_template(todolist_template_slug, project_id, step_slug)
         todos = Todo.objects.filter(
-            task__project__id=int(project.id)).order_by('-created_at')
+            project__id=project_id).order_by('-created_at')
```
Comment (marcua) [inline orchestra/todos/views.py]: why don't we have to call int() on project id anymore? is the serializer doing the cleaning for us?

File: orchestra/migrations/0087_todo_task_data_migration.py [hunk 19]
```diff
@@ -0,0 +1,28 @@
+# Generated by Django 2.2.13 on 2020-09-04 14:49
```
Comment (marcua) [inline orchestra/migrations/0087_todo_task_data_migration.py:1]: making sure we cleaned up, have you grepped for things like

`task__todo`
`todo__task__`
`task__project` [ will be noisy ]
`task__step` [ will be noisy ]
`todo.task`
`task.todos` [ still correct, but might identify broken code in its neighborhood ]

File: orchestra/migrations/0087_todo_task_data_migration.py (hunk 7 above)
Comment (jumasheff) [reply orchestra/migrations/0087_todo_task_data_migration.py:23]: This generates the following SQL–snippet:
```sql
 UPDATE "orchestra_todo"
SET    "project_id" =
       (
                  SELECT     u0."id"
                  FROM       "orchestra_project" U0
                  INNER JOIN "orchestra_task" U1
                  ON         (
                                        u0."id" = u1."project_id")
                  INNER JOIN "orchestra_todo" U2
                  ON         (
                                        u1."id" = u2."task_id")
                  WHERE      u2."id" = ("orchestra_todo"."id") limit 1),
       "step_id" =
       (
                  SELECT     u0."id"
                  FROM       "orchestra_step" u0
                  INNER JOIN "orchestra_task" u1
                  ON         (
                                        u0."id" = u1."step_id")
                  INNER JOIN "orchestra_todo" u2
                  ON         (
                                        u1."id" = u2."task_id")
                  WHERE      u2."id" = ("orchestra_todo"."id") limit 1)
WHERE  "orchestra_todo"."is_deleted" = false;args=(false,) 
```

File: orchestra/static/orchestra/todos/todo-list.directive.es6.js (hunk 9 above)
Comment (jumasheff) [reply orchestra/static/orchestra/todos/todo-list.directive.es6.js:29]:
```
13         <select class="form-control input-sm" name="todoList" id="todoList" ng-model="todoList.newTodoStepSlug">
 14           <option value="" selected>Select owner</option>
 15           <option value="{{task.step_slug}}" ng-repeat="task in todoList.possibleTasks">{{todoList.steps[task.step_slug].name}}</option>
 16         </select>
```
We set a selected `step_slug` to `newTodoStepSlug` here (in `orchestra/static/orchestra/todos/todo-list.html`).
We used to require a possible task's id to create a todo, but don't need it anymore.

File: orchestra/migrations/0087_todo_task_data_migration.py (hunk 7 above)
Comment (jumasheff) [reply orchestra/migrations/0087_todo_task_data_migration.py:23]: I intuitively tried to apply `Todo.objects.update(project=F('task__project'), step=F('task__step'))`, but it raises a Field error. But "joined F() aren't supported by design" according to this ticket: https://code.djangoproject.com/ticket/14104

File: orchestra/static/orchestra/todos/todo-list.directive.es6.js (hunk 9 above)
Comment (jumasheff) [reply orchestra/static/orchestra/todos/todo-list.directive.es6.js:29]: Hahha, I get your question now. Sorry. Can't think of any reason why we didn't get task.id from `todoList. newTodoTaskId`.

I grepped for `sendToPending`. It is not being used anywhere 🙈

File: orchestra/static/orchestra/todos/todo-list.directive.es6.js (hunk 10 above)
Comment (jumasheff) [reply orchestra/static/orchestra/todos/todo-list.directive.es6.js]: It is not being used anywhere.

File: orchestra/static/orchestra/todos/todo-list.directive.es6.js (hunk 9 above)
Comment (jumasheff) [reply orchestra/static/orchestra/todos/todo-list.directive.es6.js:29]: It was deleted in this PR (2 years ago): https://github.com/b12io/orchestra/pull/451/files#diff-0decc76f5bde80e17d109210f95a37ae

Removing the function :)

File: orchestra/static/orchestra/todos/todo-list.directive.es6.js (hunk 10 above)
Comment (jumasheff) [reply orchestra/static/orchestra/todos/todo-list.directive.es6.js]: Removed.

File: orchestra/todos/urls.py (hunk 16 above)
Comment (jumasheff) [reply orchestra/todos/urls.py]: We are 'retrofitting' `TodoViewset` to the old URLs :)
Applied rest_framework's router.

File: orchestra/todos/views.py (hunk 17 above)
Comment (jumasheff) [reply orchestra/todos/views.py:37]: Good catch! Thank you! I added `step__slug=step_slug` bit to the filter.

As a side note, validation is not available, when `data=` is not passed: https://github.com/encode/django-rest-framework/blob/e39d8410de26003421f7e644f9206833a3b9bbe0/rest_framework/serializers.py#L73

File: orchestra/todos/views.py (hunk 18 above)
Comment (jumasheff) [reply orchestra/todos/views.py]: It turs out, `Task.objects.filter(project='123')` is valid! Django cleans it itself.

File: orchestra/todos/urls.py [hunk 20]
```diff
@@ -29,3 +26,10 @@
         views.worker_task_recent_todo_qas,
         name='worker_task_recent_todo_qas'),
 ]
+
+router = routers.SimpleRouter()
```
Comment (marcua) [inline orchestra/todos/urls.py:30]: awesome! :)

just making sure this doesn't result in any newly necessary QA or exposing any new methods.

File: orchestra/utils/common_helpers.py [hunk 21]
```diff
@@ -1,5 +1,6 @@
 from rest_framework import permissions
```
Comment (marcua) [inline orchestra/utils/common_helpers.py:1]: sanity-checking: is this whole file todo_helpers and not common_helpers?

Comment (marcua) [reply orchestra/utils/common_helpers.py:1]: i don't feel strongly---i guess we can use IsAssociatedWorker elsewhere.

File: orchestra/todos/urls.py (hunk 20 above)
Comment (jumasheff) [reply orchestra/todos/urls.py:30]: Addressed. We now have this: `http_method_names = ['get', 'post', 'put']` :)

File: orchestra/utils/common_helpers.py (hunk 21 above)
Comment (jumasheff) [reply orchestra/utils/common_helpers.py:1]: Let's leave as `common_helpers`? :)

**Review by marcua — APPROVED** [review]

**Review by marcua — APPROVED** [review]

**Pull request merged**